File: CPMS12b/Pa3/Dictionary.java
//-----------------------------------------------------------------------------
// Dictionary.java
// Dictionary ADT
//-----------------------------------------------------------------------------

public class Dictionary implements DictionaryInterface{

	private class Node {
		String key;
		String value;
		Node next;

		Node(String someKey, String someValue){
			key = someKey;
			value = someValue;
			next = null;
		}
	}

	private Node head;
	private int numItems;


	public Dictionary(){
		head = new Node("","");
		head.next = null;
		numItems = 0;
	}


	// isEmpty()
	// returns true if this Dictionary is empty, false otherwise
	// pre: none
	public boolean isEmpty(){
		if(numItems == 0){
			return true;
		} else {
			return false;
		}
	}

	// size()
	// returns the number of entries in this Dictionary
	// pre: none
	public int size(){
		return numItems;
	}

	// lookup()
	// returns value associated key, or null reference if no such key exists
	// pre: none
	public String lookup(String key){
		Node N = head;
		int i = 0;
		while(N.next != null){
			if(N.next.key == key){
				return N.next.value;
			}
			N = N.next;
		}
		return null;
	}

	// insert()
	// inserts new (key,value) pair into this Dictionary
	// pre: lookup(key)==null
	public void insert(String key, String value) throws KeyCollisionException{
		if(lookup(key) != null){
			throw new KeyCollisionException("connot insert duplicate keys");
		}
		if(numItems == 0){
			Node C = new Node(key, value);
			head.next = C;
			numItems++;
		} else {
			Node N = head;
			while(N.next != null) N = N.next;
			Node C = new Node(key, value);
			N.next = C;
			numItems++;
		}
	}

	// delete()
	// deletes pair with the given key
	// pre: lookup(key)!=null
	public void delete(String key) throws KeyNotFoundException{

		if(lookup(key) == null){
			throw new KeyNotFoundException("connot delete non-existend key");
		}
		
		if(head.next.key == key){
			head.next = head.next.next;
		} else {
		Node N = new Node(key,"");
		N.next = head.next;
		while(N.next.next.key != key){
			N = N.next;
		}
		N.next.next = N.next.next.next;
		}	


		numItems--;
	}

	// makeEmpty()
	// pre: none
	public void makeEmpty(){
		head.next = null;
		numItems = 0;
	}

	// toString()
	// returns a String representation of this Dictionary
	// overrides Object's toString() method
	// pre: none
	public String toString(){
		StringBuffer sb = new StringBuffer();
		Node N = new Node("","");
		N.next = head.next;
		if(N.next == null){
			sb.append("\n");
			return new String(sb);	
		}
		for( ; N.next!=null; N.next=N.next.next) sb.append(N.next.key).append(" ").append(N.next.value).append("\n");
		return new String(sb);
	}	
}

File: CPMS12b/PA5_2/Dictionary.c
#include "dictionary.h"
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>

//-----------------------------------------------------------------------------
// Dictionary.c
// source file for the Dictionary ADT
//-----------------------------------------------------------------------------

// Dictionary
// Exported reference type
//typedef struct DictionaryObj* Dictionary;

// Private --------------------------------------------------------------------

// NodeObj
typedef struct NodeObj{
	char* key;
	char* value;
	struct NodeObj* next;
} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* key, char* value) {
	Node N = malloc(sizeof(NodeObj));
	assert(N!=NULL);
	N->key = key;
	N->value = value;
	N->next = NULL;
	return(N);
}

// freeNode()
// destructor for the Node type
void freeNode(Node *pN){
	if(pN!=NULL && *pN!=NULL){
		free(*pN);
		*pN = NULL;
	}
}

// hasgObj
typedef struct DictionaryObj{
	Node *table;
	int numItems;
}  DictionaryObj;

// Private Hash Functions -------------------------------------------------------

const int tableSize = 101;

// rotate_left()
// rotate the bits in an unsigned int
unsigned int rotate_left(unsigned int value, int shift) {
   int sizeInBits = 8*sizeof(unsigned int);
   shift = shift & (sizeInBits - 1);
   if ( shift == 0 )
      return value;
   return (value << shift) | (value >> (sizeInBits - shift));
}

// pre_hash()
// turn a string into an unsigned int
unsigned int pre_hash(char* input) { 
   unsigned int result = 0xBAE86554;
   while (*input) { 
      result ^= *input++;
      result = rotate_left(result, 5);
   }
   return result;
}

// hash()
// turns a string into an int in the range 0 to tableSize-1
int hash(char* key){
   return pre_hash(key)%tableSize;
}

// showBits()
// print out the bits in an unsigned int
void showBits(unsigned int x) {
   int i;
   int sizeInBits = 8*sizeof(unsigned int);
   char symbol[2] = {'0','1'};
   char* binary = malloc(sizeInBits + 1);

   memset(binary, 0, sizeInBits + 1);

   for (i=0; i<sizeInBits; i++) {
       binary[sizeInBits-i-1] = symbol[(x>>i) & 0x01];
   }
   
   printf("%s\n", binary);
   free(binary);
}


// Public -----------------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void)
{
	Dictionary thisDict = (Dictionary)malloc(sizeof(DictionaryObj));
	thisDict->table = (Node*)calloc(tableSize, sizeof(NodeObj));
	assert(thisDict!=NULL);
	int i = 0;
	for(i=0; i < tableSize; i++)
		assert(thisDict->table[i]!=NULL);
	thisDict->numItems = 0;
	return thisDict;
}

// freeDictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD)
{
	makeEmpty(pD);
	free(pD);
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D)
{
	if(D->numItems == 0) return 1;
	return 0;
}

// size()
// returns the number of (key, value) pairs in D
// pre: none
int size(Dictionary D)
{
	return D->numItems;
}

// lookup()
// returns the value v such that (k, v) is in D, or returns NULL if no 
// such value v exists.
// pre: none
char* lookup(Dictionary D, char* k)
{
	int index = hash(k);
	Node pN = D->table[index];
	while(pN!=NULL){
		if(strcmp(pN->key, k) == 0){
			return pN->value;
		} else {
			pN = pN->next;
		}
	}
	return NULL;
}

// insert()
// inserts new (key,value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v)
{
	int index = hash(k);
	Node pN = D->table[index];
	while(pN!=NULL){
		pN = pN->next;
	}
	pN = newNode(k, v);
	D->numItems++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k)
{
	int index = hash(k);
	Node pN = D->table[index];
	if(strcmp(pN->key, k) == 0){
		D->table[index] = D->table[index]->next;
		free(pN);
		D->numItems--;
	}
	while(pN->next!=NULL){
		if(strcmp(pN->next->key, k) == 0){
			Node temp = pN->next;
			pN->next = pN->next->next;
			free(temp);
			D->numItems--;
		} else {
			pN = pN->next;
		}
	}
}

// makeEmpty()
// re-sets D to the empty state.
// pre: none
void makeEmpty(Dictionary D)
{
	int i = 0;
	for(i=0; i<tableSize; i++){
		Node pD = D->table[i];
		while(pD!=NULL){
			Node pN = pD->next;
			free(pD);
			pD = pN;
		}
	}
	D->numItems = 0;
}

// printDictionary()
// pre: none
// prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D)
{
	int i = 0;
	for(i=0; i<tableSize; i++){
		Node pN = D->table[i];
		while(pN!=NULL){
			fprintf(out, "key: %s value: %s", D->table[i]->key, D->table[i]->value);
			pN = pN->next;
		}
	}
}

